Source/MQTTnet/Internal/AsyncEvent.cs

```diff
@@ -13,7 +14,16 @@ public sealed class AsyncEvent<TEventArgs> where TEventArgs : EventArgs
     {
         readonly List<AsyncEventInvocator<TEventArgs>> _handlers = new List<AsyncEventInvocator<TEventArgs>>();
 
-        public bool HasHandlers => _handlers.Count > 0;
+        ICollection<AsyncEventInvocator<TEventArgs>> _handlersForInvoke;
```

Comment:
Is this field necessary or can `_handlers` be used instead?
It looks quite redundant (and allocating) to me -- it's basically just `_handlers.ToList()`, so `_handlers` could be used directly?

Reply:
My idea is that the second list is build when changing the handlers inside a lock. To avoid the locking in the read access I decided to create an immutable (not really immutable but it never gets changes after initial creation) copy which is used for read access. Since this involves only a reference swap there is no need for a lock.

Reply:
👍🏻 

Is adding / removing done quite often?
Question is because of the allocation of the "immutable" list, and if it makes sense to avoid that repeated allocation in some way.
If it's not done often, the code as is, is fine and clean to read.

Reply:
I don't know how often this might be done. This change comes from another pull request where Rx is used together with this library which will add and remove handlers from different threads. My goal was to keep reading the handlers as fast as possible because I assume that changing the handlers is not done very often in most use cases.